x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/QueryRule.java

```diff
@@ -283,29 +287,39 @@ public AppliedQueryRules applyRule(AppliedQueryRules appliedRules, Map<String, O
 
         List<String> matchingPinnedIds = new ArrayList<>();
         List<PinnedQueryBuilder.Item> matchingPinnedDocs = new ArrayList<>();
+        Boolean isRuleMatch = null;
 
+        // All specified criteria in a rule must match for the rule to be applied
         for (QueryRuleCriteria criterion : criteria) {
             for (String match : matchCriteria.keySet()) {
-                final String matchValue = matchCriteria.get(match).toString();
-                if (criterion.criteriaMetadata().equals(match) && criterion.isMatch(matchValue)) {
-                    if (actions.containsKey(IDS_FIELD.getPreferredName())) {
-                        matchingPinnedIds.addAll((List<String>) actions.get(IDS_FIELD.getPreferredName()));
-                    } else if (actions.containsKey(DOCS_FIELD.getPreferredName())) {
-                        List<Map<String, String>> docsToPin = (List<Map<String, String>>) actions.get(DOCS_FIELD.getPreferredName());
-                        List<PinnedQueryBuilder.Item> items = docsToPin.stream()
-                            .map(
-                                map -> new PinnedQueryBuilder.Item(
-                                    map.get(INDEX_FIELD.getPreferredName()),
-                                    map.get(PinnedQueryBuilder.Item.ID_FIELD.getPreferredName())
-                                )
-                            )
-                            .toList();
-                        matchingPinnedDocs.addAll(items);
-                    }
+                final Object matchValue = matchCriteria.get(match);
+                final QueryRuleCriteria.CriteriaType criteriaType = criterion.criteriaType();
+                final String criteriaMetadata = criterion.criteriaMetadata();
+
+                if (criteriaType == QueryRuleCriteria.CriteriaType.GLOBAL || (criteriaMetadata != null && criteriaMetadata.equals(match))) {
+                    boolean singleCriterionMatches = criterion.isMatch(matchValue, criteriaType);
+                    isRuleMatch = (isRuleMatch == null) ? singleCriterionMatches : isRuleMatch & singleCriterionMatches;
```

Comment:
```suggestion
                    isRuleMatch = (isRuleMatch == null) ? singleCriterionMatches : isRuleMatch && singleCriterionMatches;
```

Reply:
I think we should use the double `&&` here, or is there a reason to not short circuit the AND operation?

Reply:
You're right, that was a typo, thanks!